userapi/internal/device_list_update.go

```diff
@@ -405,24 +418,40 @@ func (u *DeviceListUpdater) worker(ch chan spec.ServerName, workerID int) {
 	go func() {
 		var serversToRetry []spec.ServerName
 		for {
+			deviceListUpdaterServersRetrying.With(prometheus.Labels{"worker_id": strconv.Itoa(workerID)}).Set(float64(len(retries)))
+			time.Sleep(time.Second * 2)
+
+			// The channel is at capacity, don't try to send more work
+			if len(ch) == cap(ch) {
+				continue
+			}
 			serversToRetry = serversToRetry[:0] // reuse memory
-			time.Sleep(time.Second)
+
+			// -2, so we have space for incoming device list updates over federation
+			maxServers := (cap(ch) - len(ch)) - 2
+			if maxServers <= 0 {
```

Comment:
This check is also effectively doing `if len(ch) == cap(ch)` because if they are `==` then `N-N-2 <= 0`, maybe just have this check and drop the earlier one?